// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: indykite/ingest/v1beta3/model.proto

package ingestv1beta3

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on Record with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Record) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Record with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in RecordMultiError, or nil if none found.
func (m *Record) ValidateAll() error {
	return m.validate(true)
}

func (m *Record) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if utf8.RuneCountInString(m.GetId()) > 256 {
		err := RecordValidationError{
			field:  "Id",
			reason: "value length must be at most 256 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	oneofOperationPresent := false
	switch v := m.Operation.(type) {
	case *Record_Upsert:
		if v == nil {
			err := RecordValidationError{
				field:  "Operation",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		oneofOperationPresent = true

		if all {
			switch v := interface{}(m.GetUpsert()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, RecordValidationError{
						field:  "Upsert",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, RecordValidationError{
						field:  "Upsert",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetUpsert()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return RecordValidationError{
					field:  "Upsert",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *Record_Delete:
		if v == nil {
			err := RecordValidationError{
				field:  "Operation",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		oneofOperationPresent = true

		if all {
			switch v := interface{}(m.GetDelete()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, RecordValidationError{
						field:  "Delete",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, RecordValidationError{
						field:  "Delete",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetDelete()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return RecordValidationError{
					field:  "Delete",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	default:
		_ = v // ensures v is used
	}
	if !oneofOperationPresent {
		err := RecordValidationError{
			field:  "Operation",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return RecordMultiError(errors)
	}

	return nil
}

// RecordMultiError is an error wrapping multiple validation errors returned by
// Record.ValidateAll() if the designated constraints aren't met.
type RecordMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RecordMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RecordMultiError) AllErrors() []error { return m }

// RecordValidationError is the validation error returned by Record.Validate if
// the designated constraints aren't met.
type RecordValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RecordValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RecordValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RecordValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RecordValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RecordValidationError) ErrorName() string { return "RecordValidationError" }

// Error satisfies the builtin error interface
func (e RecordValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRecord.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RecordValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RecordValidationError{}

// Validate checks the field values on UpsertData with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *UpsertData) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpsertData with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in UpsertDataMultiError, or
// nil if none found.
func (m *UpsertData) ValidateAll() error {
	return m.validate(true)
}

func (m *UpsertData) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	oneofDataPresent := false
	switch v := m.Data.(type) {
	case *UpsertData_Node:
		if v == nil {
			err := UpsertDataValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		oneofDataPresent = true

		if all {
			switch v := interface{}(m.GetNode()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, UpsertDataValidationError{
						field:  "Node",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, UpsertDataValidationError{
						field:  "Node",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetNode()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return UpsertDataValidationError{
					field:  "Node",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *UpsertData_Relationship:
		if v == nil {
			err := UpsertDataValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		oneofDataPresent = true

		if all {
			switch v := interface{}(m.GetRelationship()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, UpsertDataValidationError{
						field:  "Relationship",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, UpsertDataValidationError{
						field:  "Relationship",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetRelationship()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return UpsertDataValidationError{
					field:  "Relationship",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	default:
		_ = v // ensures v is used
	}
	if !oneofDataPresent {
		err := UpsertDataValidationError{
			field:  "Data",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return UpsertDataMultiError(errors)
	}

	return nil
}

// UpsertDataMultiError is an error wrapping multiple validation errors
// returned by UpsertData.ValidateAll() if the designated constraints aren't met.
type UpsertDataMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpsertDataMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpsertDataMultiError) AllErrors() []error { return m }

// UpsertDataValidationError is the validation error returned by
// UpsertData.Validate if the designated constraints aren't met.
type UpsertDataValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpsertDataValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpsertDataValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpsertDataValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpsertDataValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpsertDataValidationError) ErrorName() string { return "UpsertDataValidationError" }

// Error satisfies the builtin error interface
func (e UpsertDataValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpsertData.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpsertDataValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpsertDataValidationError{}

// Validate checks the field values on DeleteData with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DeleteData) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteData with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DeleteDataMultiError, or
// nil if none found.
func (m *DeleteData) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteData) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	oneofDataPresent := false
	switch v := m.Data.(type) {
	case *DeleteData_Node:
		if v == nil {
			err := DeleteDataValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		oneofDataPresent = true

		if all {
			switch v := interface{}(m.GetNode()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, DeleteDataValidationError{
						field:  "Node",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, DeleteDataValidationError{
						field:  "Node",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetNode()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return DeleteDataValidationError{
					field:  "Node",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *DeleteData_Relationship:
		if v == nil {
			err := DeleteDataValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		oneofDataPresent = true

		if all {
			switch v := interface{}(m.GetRelationship()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, DeleteDataValidationError{
						field:  "Relationship",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, DeleteDataValidationError{
						field:  "Relationship",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetRelationship()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return DeleteDataValidationError{
					field:  "Relationship",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *DeleteData_NodeProperty:
		if v == nil {
			err := DeleteDataValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		oneofDataPresent = true

		if all {
			switch v := interface{}(m.GetNodeProperty()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, DeleteDataValidationError{
						field:  "NodeProperty",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, DeleteDataValidationError{
						field:  "NodeProperty",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetNodeProperty()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return DeleteDataValidationError{
					field:  "NodeProperty",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *DeleteData_RelationshipProperty:
		if v == nil {
			err := DeleteDataValidationError{
				field:  "Data",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		oneofDataPresent = true

		if all {
			switch v := interface{}(m.GetRelationshipProperty()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, DeleteDataValidationError{
						field:  "RelationshipProperty",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, DeleteDataValidationError{
						field:  "RelationshipProperty",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetRelationshipProperty()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return DeleteDataValidationError{
					field:  "RelationshipProperty",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	default:
		_ = v // ensures v is used
	}
	if !oneofDataPresent {
		err := DeleteDataValidationError{
			field:  "Data",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return DeleteDataMultiError(errors)
	}

	return nil
}

// DeleteDataMultiError is an error wrapping multiple validation errors
// returned by DeleteData.ValidateAll() if the designated constraints aren't met.
type DeleteDataMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteDataMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteDataMultiError) AllErrors() []error { return m }

// DeleteDataValidationError is the validation error returned by
// DeleteData.Validate if the designated constraints aren't met.
type DeleteDataValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteDataValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteDataValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteDataValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteDataValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteDataValidationError) ErrorName() string { return "DeleteDataValidationError" }

// Error satisfies the builtin error interface
func (e DeleteDataValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteData.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteDataValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteDataValidationError{}

// Validate checks the field values on Relationship with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Relationship) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Relationship with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in RelationshipMultiError, or
// nil if none found.
func (m *Relationship) ValidateAll() error {
	return m.validate(true)
}

func (m *Relationship) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetSource() == nil {
		err := RelationshipValidationError{
			field:  "Source",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetSource()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, RelationshipValidationError{
					field:  "Source",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, RelationshipValidationError{
					field:  "Source",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetSource()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return RelationshipValidationError{
				field:  "Source",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if m.GetTarget() == nil {
		err := RelationshipValidationError{
			field:  "Target",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetTarget()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, RelationshipValidationError{
					field:  "Target",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, RelationshipValidationError{
					field:  "Target",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTarget()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return RelationshipValidationError{
				field:  "Target",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if utf8.RuneCountInString(m.GetType()) > 128 {
		err := RelationshipValidationError{
			field:  "Type",
			reason: "value length must be at most 128 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if !_Relationship_Type_Pattern.MatchString(m.GetType()) {
		err := RelationshipValidationError{
			field:  "Type",
			reason: "value does not match regex pattern \"^[A-Z]+(?:_[A-Z]+)*$\"",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(m.GetProperties()) > 50 {
		err := RelationshipValidationError{
			field:  "Properties",
			reason: "value must contain no more than 50 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	for idx, item := range m.GetProperties() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, RelationshipValidationError{
						field:  fmt.Sprintf("Properties[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, RelationshipValidationError{
						field:  fmt.Sprintf("Properties[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return RelationshipValidationError{
					field:  fmt.Sprintf("Properties[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return RelationshipMultiError(errors)
	}

	return nil
}

// RelationshipMultiError is an error wrapping multiple validation errors
// returned by Relationship.ValidateAll() if the designated constraints aren't met.
type RelationshipMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RelationshipMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RelationshipMultiError) AllErrors() []error { return m }

// RelationshipValidationError is the validation error returned by
// Relationship.Validate if the designated constraints aren't met.
type RelationshipValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RelationshipValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RelationshipValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RelationshipValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RelationshipValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RelationshipValidationError) ErrorName() string { return "RelationshipValidationError" }

// Error satisfies the builtin error interface
func (e RelationshipValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRelationship.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RelationshipValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RelationshipValidationError{}

var _Relationship_Type_Pattern = regexp.MustCompile("^[A-Z]+(?:_[A-Z]+)*$")

// Validate checks the field values on NodeMatch with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *NodeMatch) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on NodeMatch with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in NodeMatchMultiError, or nil
// if none found.
func (m *NodeMatch) ValidateAll() error {
	return m.validate(true)
}

func (m *NodeMatch) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if l := utf8.RuneCountInString(m.GetExternalId()); l < 1 || l > 256 {
		err := NodeMatchValidationError{
			field:  "ExternalId",
			reason: "value length must be between 1 and 256 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetType()) > 64 {
		err := NodeMatchValidationError{
			field:  "Type",
			reason: "value length must be at most 64 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if !_NodeMatch_Type_Pattern.MatchString(m.GetType()) {
		err := NodeMatchValidationError{
			field:  "Type",
			reason: "value does not match regex pattern \"^([A-Z][a-z]+)+$\"",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return NodeMatchMultiError(errors)
	}

	return nil
}

// NodeMatchMultiError is an error wrapping multiple validation errors returned
// by NodeMatch.ValidateAll() if the designated constraints aren't met.
type NodeMatchMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m NodeMatchMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m NodeMatchMultiError) AllErrors() []error { return m }

// NodeMatchValidationError is the validation error returned by
// NodeMatch.Validate if the designated constraints aren't met.
type NodeMatchValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NodeMatchValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NodeMatchValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NodeMatchValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NodeMatchValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NodeMatchValidationError) ErrorName() string { return "NodeMatchValidationError" }

// Error satisfies the builtin error interface
func (e NodeMatchValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNodeMatch.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NodeMatchValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NodeMatchValidationError{}

var _NodeMatch_Type_Pattern = regexp.MustCompile("^([A-Z][a-z]+)+$")

// Validate checks the field values on RecordError with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *RecordError) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on RecordError with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in RecordErrorMultiError, or
// nil if none found.
func (m *RecordError) ValidateAll() error {
	return m.validate(true)
}

func (m *RecordError) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Error

	if len(errors) > 0 {
		return RecordErrorMultiError(errors)
	}

	return nil
}

// RecordErrorMultiError is an error wrapping multiple validation errors
// returned by RecordError.ValidateAll() if the designated constraints aren't met.
type RecordErrorMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RecordErrorMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RecordErrorMultiError) AllErrors() []error { return m }

// RecordErrorValidationError is the validation error returned by
// RecordError.Validate if the designated constraints aren't met.
type RecordErrorValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RecordErrorValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RecordErrorValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RecordErrorValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RecordErrorValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RecordErrorValidationError) ErrorName() string { return "RecordErrorValidationError" }

// Error satisfies the builtin error interface
func (e RecordErrorValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRecordError.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RecordErrorValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RecordErrorValidationError{}

// Validate checks the field values on Info with the rules defined in the proto
// definition for this message. If any rules are violated, the first error
// encountered is returned, or nil if there are no violations.
func (m *Info) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Info with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in InfoMultiError, or nil if none found.
func (m *Info) ValidateAll() error {
	return m.validate(true)
}

func (m *Info) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetChanges() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, InfoValidationError{
						field:  fmt.Sprintf("Changes[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, InfoValidationError{
						field:  fmt.Sprintf("Changes[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return InfoValidationError{
					field:  fmt.Sprintf("Changes[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return InfoMultiError(errors)
	}

	return nil
}

// InfoMultiError is an error wrapping multiple validation errors returned by
// Info.ValidateAll() if the designated constraints aren't met.
type InfoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m InfoMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m InfoMultiError) AllErrors() []error { return m }

// InfoValidationError is the validation error returned by Info.Validate if the
// designated constraints aren't met.
type InfoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e InfoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e InfoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e InfoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e InfoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e InfoValidationError) ErrorName() string { return "InfoValidationError" }

// Error satisfies the builtin error interface
func (e InfoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sInfo.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = InfoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = InfoValidationError{}

// Validate checks the field values on Change with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Change) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Change with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in ChangeMultiError, or nil if none found.
func (m *Change) ValidateAll() error {
	return m.validate(true)
}

func (m *Change) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for DataType

	if len(errors) > 0 {
		return ChangeMultiError(errors)
	}

	return nil
}

// ChangeMultiError is an error wrapping multiple validation errors returned by
// Change.ValidateAll() if the designated constraints aren't met.
type ChangeMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ChangeMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ChangeMultiError) AllErrors() []error { return m }

// ChangeValidationError is the validation error returned by Change.Validate if
// the designated constraints aren't met.
type ChangeValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ChangeValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ChangeValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ChangeValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ChangeValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ChangeValidationError) ErrorName() string { return "ChangeValidationError" }

// Error satisfies the builtin error interface
func (e ChangeValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sChange.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ChangeValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ChangeValidationError{}

// Validate checks the field values on DeleteData_NodePropertyMatch with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteData_NodePropertyMatch) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteData_NodePropertyMatch with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeleteData_NodePropertyMatchMultiError, or nil if none found.
func (m *DeleteData_NodePropertyMatch) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteData_NodePropertyMatch) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetMatch() == nil {
		err := DeleteData_NodePropertyMatchValidationError{
			field:  "Match",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetMatch()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DeleteData_NodePropertyMatchValidationError{
					field:  "Match",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DeleteData_NodePropertyMatchValidationError{
					field:  "Match",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetMatch()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DeleteData_NodePropertyMatchValidationError{
				field:  "Match",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if m.GetPropertyType() != "" {

		if len(m.GetPropertyType()) > 256 {
			err := DeleteData_NodePropertyMatchValidationError{
				field:  "PropertyType",
				reason: "value length must be at most 256 bytes",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if !_DeleteData_NodePropertyMatch_PropertyType_Pattern.MatchString(m.GetPropertyType()) {
			err := DeleteData_NodePropertyMatchValidationError{
				field:  "PropertyType",
				reason: "value does not match regex pattern \"^[a-zA-Z_][a-zA-Z0-9_]+$\"",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if len(errors) > 0 {
		return DeleteData_NodePropertyMatchMultiError(errors)
	}

	return nil
}

// DeleteData_NodePropertyMatchMultiError is an error wrapping multiple
// validation errors returned by DeleteData_NodePropertyMatch.ValidateAll() if
// the designated constraints aren't met.
type DeleteData_NodePropertyMatchMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteData_NodePropertyMatchMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteData_NodePropertyMatchMultiError) AllErrors() []error { return m }

// DeleteData_NodePropertyMatchValidationError is the validation error returned
// by DeleteData_NodePropertyMatch.Validate if the designated constraints
// aren't met.
type DeleteData_NodePropertyMatchValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteData_NodePropertyMatchValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteData_NodePropertyMatchValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteData_NodePropertyMatchValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteData_NodePropertyMatchValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteData_NodePropertyMatchValidationError) ErrorName() string {
	return "DeleteData_NodePropertyMatchValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteData_NodePropertyMatchValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteData_NodePropertyMatch.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteData_NodePropertyMatchValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteData_NodePropertyMatchValidationError{}

var _DeleteData_NodePropertyMatch_PropertyType_Pattern = regexp.MustCompile("^[a-zA-Z_][a-zA-Z0-9_]+$")

// Validate checks the field values on DeleteData_RelationshipPropertyMatch
// with the rules defined in the proto definition for this message. If any
// rules are violated, the first error encountered is returned, or nil if
// there are no violations.
func (m *DeleteData_RelationshipPropertyMatch) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteData_RelationshipPropertyMatch
// with the rules defined in the proto definition for this message. If any
// rules are violated, the result is a list of violation errors wrapped in
// DeleteData_RelationshipPropertyMatchMultiError, or nil if none found.
func (m *DeleteData_RelationshipPropertyMatch) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteData_RelationshipPropertyMatch) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetSource() == nil {
		err := DeleteData_RelationshipPropertyMatchValidationError{
			field:  "Source",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetSource()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DeleteData_RelationshipPropertyMatchValidationError{
					field:  "Source",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DeleteData_RelationshipPropertyMatchValidationError{
					field:  "Source",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetSource()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DeleteData_RelationshipPropertyMatchValidationError{
				field:  "Source",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if m.GetTarget() == nil {
		err := DeleteData_RelationshipPropertyMatchValidationError{
			field:  "Target",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetTarget()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DeleteData_RelationshipPropertyMatchValidationError{
					field:  "Target",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DeleteData_RelationshipPropertyMatchValidationError{
					field:  "Target",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTarget()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DeleteData_RelationshipPropertyMatchValidationError{
				field:  "Target",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if utf8.RuneCountInString(m.GetType()) > 128 {
		err := DeleteData_RelationshipPropertyMatchValidationError{
			field:  "Type",
			reason: "value length must be at most 128 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if !_DeleteData_RelationshipPropertyMatch_Type_Pattern.MatchString(m.GetType()) {
		err := DeleteData_RelationshipPropertyMatchValidationError{
			field:  "Type",
			reason: "value does not match regex pattern \"^[A-Z]+(?:_[A-Z]+)*$\"",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.GetPropertyType() != "" {

		if len(m.GetPropertyType()) > 256 {
			err := DeleteData_RelationshipPropertyMatchValidationError{
				field:  "PropertyType",
				reason: "value length must be at most 256 bytes",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if !_DeleteData_RelationshipPropertyMatch_PropertyType_Pattern.MatchString(m.GetPropertyType()) {
			err := DeleteData_RelationshipPropertyMatchValidationError{
				field:  "PropertyType",
				reason: "value does not match regex pattern \"^[a-zA-Z_][a-zA-Z0-9_]+$\"",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if len(errors) > 0 {
		return DeleteData_RelationshipPropertyMatchMultiError(errors)
	}

	return nil
}

// DeleteData_RelationshipPropertyMatchMultiError is an error wrapping multiple
// validation errors returned by
// DeleteData_RelationshipPropertyMatch.ValidateAll() if the designated
// constraints aren't met.
type DeleteData_RelationshipPropertyMatchMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteData_RelationshipPropertyMatchMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteData_RelationshipPropertyMatchMultiError) AllErrors() []error { return m }

// DeleteData_RelationshipPropertyMatchValidationError is the validation error
// returned by DeleteData_RelationshipPropertyMatch.Validate if the designated
// constraints aren't met.
type DeleteData_RelationshipPropertyMatchValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteData_RelationshipPropertyMatchValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteData_RelationshipPropertyMatchValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteData_RelationshipPropertyMatchValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteData_RelationshipPropertyMatchValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteData_RelationshipPropertyMatchValidationError) ErrorName() string {
	return "DeleteData_RelationshipPropertyMatchValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteData_RelationshipPropertyMatchValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteData_RelationshipPropertyMatch.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteData_RelationshipPropertyMatchValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteData_RelationshipPropertyMatchValidationError{}

var _DeleteData_RelationshipPropertyMatch_Type_Pattern = regexp.MustCompile("^[A-Z]+(?:_[A-Z]+)*$")

var _DeleteData_RelationshipPropertyMatch_PropertyType_Pattern = regexp.MustCompile("^[a-zA-Z_][a-zA-Z0-9_]+$")
